climate_index_collection/reductions.py

```diff
@@ -302,9 +302,19 @@ def area_mean_weighted(
     lon_east = lon_east % 360
 
     # coords
-    mask = (
-        (lat >= lat_south) & (lat <= lat_north) & (lon >= lon_west) & (lon <= lon_east)
-    )
+    if lon_west <= lon_east:
+        mask = (
+            (lat >= lat_south)
+            & (lat <= lat_north)
+            & (lon >= lon_west)
+            & (lon <= lon_east)
+        )
+    else:
+        mask = (lat == 0) & (lon == 0) + False
+        mask_west = (lat >= lat_south) & (lat <= lat_north) & (lon >= lon_west)
+        mask_east = (lat >= lat_south) & (lat <= lat_north) & (lon <= lon_east)
+        mask_west
+        mask = mask + np.ma.mask_or(mask_west, mask_east)
```

Comment:
I think we should do three things:
1. Create a separate function for calculating the mask. This helps with 2. below.

2. Write a test covering the issue that's meant to be solved here.

3. Maybe simplify the logic? The problem we have here can be solved by switching the base interval of the lons from `[0, 360)` to `[-180, 180)` if in the former base interval, `lat_west` is larger than `lat_east`?

Reply:
Ad 3:
```python
# normal case
lons = [0, 30, 60, 90, 120, 150, 180, 210, 240, 270, 300, 330]
lon_west = 50
lon_east = 220

print(list(filter(lambda l: (l >= lon_west) & (l <= lon_east), lons)))
```
```
[60, 90, 120, 150, 180, 210]
```
```python
# cross-greenwich case
lons = [0, 30, 60, 90, 120, 150, 180, 210, 240, 270, 300, 330]
lon_west = 220
lon_east = 50

def to_pm_180(lon):
    return lon - 360 * ((lon % 360) > 180)

lons = list(map(to_pm_180, lons))
lon_west = to_pm_180(lon_west)
lon_east = to_pm_180(lon_east)

print(list(filter(lambda l: (l >= lon_west) & (l <= lon_east), lons)))
```
```
[0, 30, -120, -90, -60, -30]
```

Reply:
- Regarding point 1 and 2 I certainly agree with you.

- On 3: Hope I'm not wrong but won't we then still have the issue of an 'artificial' prime meridian at 180°W which we still can't cross with our selection method.

Reply:
Re 3: Yes. The full logic would be
```python
def calc_mask(lon, lat, lon_w, lon_e, ...):
    if (lon_w % 360.0) <= (lon_e % 360.0):  # normal case
        lon = lon % 360.0
        lon_w = lon_w % 360.0
        lon_e = lon_e % 360.0
    else:  # shifted case
        lon = to_pm_180(lon)
        lon_w = to_pm_180(lon_w)
        lon_e = to_pm_180(lon_e)

    # then go on to construct the mask in the same way for both.

```

Reply:
(Sorry, I should have been more explicit about this.)

Reply:
Ok I'm definitely fine with 1 and 2.
But for 3:
I think we still aren't covering all possibilities.
Assuming we want to select an area between 60°W and 20°W, therefore crossing 0°W and 180°W, we won't be able to do so based on your idea.
This is because we just created a new 'artificial' prime meridian which we again can not cross.
Might be that I'm wrong here.

Reply:
I tried to go for the simplification in 9d33e5b3ecd2d3e155ad985f19cf65e01b02069d
At the moment I don't have the time to finish the testing I started to implement in the new .ipynb .

Reply:
I can take over from here. Thanks for starting this. Looks like it's almost done.